controllers/mariabackup_controller.go

```diff
@@ -0,0 +1,218 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MariaBackupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var backup mariadbv1alpha1.MariaBackup
+	if err := r.Get(ctx, req.NamespacedName, &backup); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	mariaDb, err := r.RefResolver.MariaDB(ctx, &backup.Spec.MariaDBRef, backup.Namespace)
+	if err != nil {
+		var mariaDbErr *multierror.Error
+		mariaDbErr = multierror.Append(mariaDbErr, err)
+
+		err = r.patchStatus(ctx, &backup, r.ConditionComplete.PatcherRefResolver(err, mariaDb))
+		mariaDbErr = multierror.Append(mariaDbErr, err)
+
+		return ctrl.Result{}, fmt.Errorf("error getting MariaDB: %v", mariaDbErr)
+	}
+
+	if backup.Spec.MariaDBRef.WaitForIt && !mariaDb.IsReady() {
+		if err := r.patchStatus(ctx, &backup, r.ConditionComplete.PatcherFailed("MariaDB not ready")); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error patching Backup: %v", err)
+		}
+		return ctrl.Result{}, errors.New("MariaDB not ready")
+	}
+
+	var batchErr *multierror.Error
+	err = r.BatchReconciler.Reconcile(ctx, &backup, mariaDb)
+	batchErr = multierror.Append(batchErr, err)
+
+	patcher, err := r.patcher(ctx, err, req.NamespacedName, &backup)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		return ctrl.Result{}, fmt.Errorf("error getting patcher for Backup: %v", err)
+	}
+
+	err = r.patchStatus(ctx, &backup, patcher)
+	batchErr = multierror.Append(batchErr, err)
+
+	if err := batchErr.ErrorOrNil(); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error creating Job: %v", err)
+	}
+
+	podList := &corev1.PodList{}
+
+	// Create a label selector to filter pods with a specific label
+	labelSelector := labels.SelectorFromSet(labels.Set(map[string]string{
+		"backup-ref": backup.Name,
+	}))
+
+	// Create a ListOptions with the label selector
+	listOptions := &client.ListOptions{
+		LabelSelector: labelSelector,
+	}
+
+	max_attempts := 0
+	for len(podList.Items) == 0 && max_attempts < 10 {
+		// List pods in the specified namespace with the label selector
+		err = r.Client.List(ctx, podList, listOptions)
+		if err != nil {
+			// Handle the error
+			return ctrl.Result{}, fmt.Errorf("error getting job: %v", err)
+		}
+		time.Sleep(3 * time.Second)
+		max_attempts++
+	}
+
+	var podIP string = ""
+	// Iterate through the list of pods
+
+	for _, pod := range podList.Items {
+
+		if pod.Status.Phase == "Running" {
+			podIP = pod.Status.PodIP
+			break
+		} else if pod.Status.Phase == "Succeeded" {
+			podIP = ""
+			break
+		}
+	}
+
+	if podIP != "" {
+		// Start mariadb-backup on the primary pod
+		// TODO: make the backup source configurable primary or replicas
+		// TODO: add s3 configuration parameters to get backup to s3 directly
```

Comment:
S3 backup storage is in our roadmap already and something that our community is looking forward to have: https://github.com/mariadb-operator/mariadb-operator/issues/6

If you feel like the PR gets too big feel free to open a new one, or we can take care of it as a separate issue, I leave it up to you. We probably would need to generalize it, not only for physical but also for logical backups.

Reply:
well, my initial idea was using the same logic that we have in xtrabackup + xbcloud [here](https://docs.percona.com/percona-xtrabackup/8.0/xbcloud-swift.html) that percona has built-in support to get backup directly to s3 and also restore as well. However, it seems there is no such utility like mariadbcloud utility for mariadbbackup. I need to think about this a bit more.

Just as an idea, I could use xtrabackup [image](https://hub.docker.com/r/percona/percona-xtrabackup) in the job pod and it would receive mariabackup stream through socat and upload to bucket with xbcloud like mariadbbackup + xbcloud. As mariadbbackup is ported from xtrabackup hopefully they are compatible.

Reply:
Regarding S3, we have a builtin mechanism based of the `mariadb-operator` image:
- https://github.com/mariadb-operator/mariadb-operator/tree/main/cmd/backup
So we don't need to rely on xbcloud